mlir/lib/Dialect/MIOpen/Transforms/BlockwiseGemmToThreadwise.cpp

```diff
@@ -112,42 +112,80 @@ struct BlockwiseGemmRewritePattern
-                            << "NPerThread: " << nPerThread << "\n");
-
-    TopDownTMBuilder strideLDSBufferA(b,
-                                      {"k", "mRepeat", "mPerThread", "kpack"},
-                                      {k, mRepeat, m / mRepeat, kPack}, loc);
-    strideLDSBufferA.passThrough("k");
-    strideLDSBufferA.embed("m", 1, m, {"mRepeat", "mPerThread"},
-                           {mRepeatStride, 1});
-    strideLDSBufferA.passThrough({"kpack"}, {2}, {"kpack"});
-    TransformMapAttr strideLDSBufferAAttr = strideLDSBufferA.get();
-
-    TopDownTMBuilder strideLDSBufferB(b,
-                                      {"k", "nRepeat", "nPerThread", "kpack"},
-                                      {k, nRepeat, n / nRepeat, kPack}, loc);
-    strideLDSBufferB.passThrough("k");
-    strideLDSBufferB.embed("n", 1, n, {"nRepeat", "nPerThread"},
-                           {nRepeatStride, 1});
-    strideLDSBufferB.passThrough({"kpack"}, {2}, {"kpack"});
-    TransformMapAttr strideLDSBufferBAttr = strideLDSBufferB.get();
+    LLVM_DEBUG(llvm::dbgs()
+               << "M: " << m << "\n"
+               << "mRepeat: " << mRepeat << "\n"
+               << "mCuwavesPerBlock: " << mCuwavesPerBlock << "\n"
+               << "mThreadsPerCuwave: " << mThreadsPerCuwave << "\n"
+               << "mPerThread: " << mPerThread << "\n"
+               << "n: " << n << "\n"
+               << "nRepeat: " << nRepeat << "\n"
+               << "nCuwavesPerBlock: " << nCuwavesPerBlock << "\n"
+               << "nThreadsPerCuwave: " << nThreadsPerCuwave << "\n"
+               << "nPerThread: " << nPerThread << "\n");
+
+    auto ldsTidSplitter = [&](StringRef repeatName, int64_t repeatLen,
+                              StringRef perThreadName,
+                              int64_t perThreadLen) -> TopDownTMBuilder {
+      TopDownTMBuilder splitTidForLDS(
+          b, {"k", repeatName, "tid", perThreadName, "kpack"},
+          {k, repeatLen, blockSize, perThreadLen, kPack}, loc);
+      splitTidForLDS.passThrough({"k", repeatName});
+      splitTidForLDS.merge({"m_cuwaves", "n_cuwaves", "m_cuwave", "n_cuwave"},
+                           {2, 3, 4, 5}, "tid",
+                           {mCuwavesPerBlock, nCuwavesPerBlock,
+                            mThreadsPerCuwave, nThreadsPerCuwave},
+                           /*isUnfold=*/true);
+      splitTidForLDS.passThrough({perThreadName, "kpack"}, {6, 7},
+                                 {perThreadName, "kpack"});
+      return splitTidForLDS;
+    };
+
+    TopDownTMBuilder splitTidA =
+        ldsTidSplitter("m_repeat", mRepeat, "m_thread", mPerThread);
+    TransformMapAttr splitTidAAttr = splitTidA.get();
+    auto toLdsIndexA = TopDownTMBuilder::below(splitTidA, splitTidAAttr);
+    toLdsIndexA.passThrough("k");
+    toLdsIndexA.unmerge(
+        "m", 1, {"m_repeat", "m_cuwaves", "m_cuwave", "m_thread"},
+        {mRepeat, mCuwavesPerBlock, mThreadsPerCuwave, mPerThread});
```

Comment:
I'm a bit lost on why you merge and then unmerge.

If the past and current `TopDownTMBuilder` are equivalent, I'd honestly think the past is more readable: I can immediately figure out the layout of `K, M, KPack` or `K, N, KPack`. Or my assumption is wrong? Did you do anything in addition to past builder?

Reply:
This isn't equivalent to the old version. In the old version, there was an implicit thread offset being passed in from `gridwise_gemm`, which is now gone. Now, we have to compute our offset ourselves, which leads to the fiddlyness. I could go further here and make this a `tid x iteration number` function that then splits into `k`, `m_repeat`, `m_thread`, and `kpack`, but since the buffer layout is already `k x [m total] x kpack`, sticking to those coordinates made sense

Reply:
Just to make sure I understand. The builder previously is:
 - `k - <m_cuwaves, m_cuwave> - kpack`

Now, it is:
 - `k - < m_repeat, m_cuwaves, m_cuwave, m_thread > - kpack`

Since the dimension size of `m_cuwaves` and `m_cuwave` are the same, does that mean you added two dimensions of 1 into the transform?

Further down into `viewA`, it looks like you have: `k - < m_repeat, tid, m_thread > - kpack`. In which `tid` dimension is equivalent to `m_cuwaves, m_cuwave`?

Sorry if I'm asking noob questions on coordinate transfer, I don't quite see all transformations add up through the way it is composed.

Reply:
Both transformations start from `k, m_repeat, tid, m_thread, kpack`
When walking our way towards the vector, we ignore `tid`.

When working towards matrix A, we split `tid` into `m_cuwaves, n_cuwaves, m_cuwave n_cuwave` and then drop the `n_*` ones in order to get a `m` coordinate. 

`m_cuwaves` and `m_cuwave` don't have the same dimension: although currently `m_cuwave` is always `4`, `m_cuwaves` is a function of `blockSize`

Reply:
And previously we had `k, m_repeat, m_thread, kpack`, where `m_thread`'s initial value was based on the function currently being computed from `m_cuwaves` and `m_cuwave`.